server/runtime/src/main/java/org/infinispan/server/Server.java

```diff
@@ -303,16 +322,79 @@ public ConfigurationInfo getConfiguration() {
       return serverConfiguration;
    }
 
-   private void shutdown() {
-      status = ComponentStatus.STOPPING;
+   @Override
+   public void serverStop() {
+      // Tell the local server to exit
+      serverStopHandler(ExitStatus.SERVER_SHUTDOWN);
+   }
+
+   @Override
+   public void serverStop(List<String> servers) {
+      for (DefaultCacheManager cacheManager : cacheManagers.values()) {
+         // Find the actual addresses of the servers
+         List<Address> targets = cacheManager.getMembers().stream()
+               .filter(a -> servers.contains(a.toString()))
+               .collect(Collectors.toList());
+         // Tell all the target servers to exit
+         sendExitStatusToServers(cacheManager.executor().filterTargets(targets), ExitStatus.SERVER_SHUTDOWN);
+      }
+   }
+
+   @Override
+   public void clusterStop() {
+      cacheManagers.values().forEach(cm -> {
+         cm.getCacheNames().forEach(name -> SecurityActions.shutdownCache(cm, name));
+         sendExitStatusToServers(cm.executor(), ExitStatus.CLUSTER_SHUTDOWN);
+      });
+   }
+
+   private void sendExitStatusToServers(ClusterExecutor clusterExecutor, ExitStatus exitStatus) {
+      CompletableFuture<Void> job = clusterExecutor.submitConsumer(new ShutdownRunnable(exitStatus), (a, i, t) -> {
+         if (t != null) {
+            log.clusteredTaskError(t);
+         }
+      });
+      job.join();
+   }
+
+   private void localShutdown(ExitStatus exitStatus) {
+      this.status = ComponentStatus.STOPPING;
+      if (exitStatus == ExitStatus.CLUSTER_SHUTDOWN) {
+         log.clusterShutdown();
+      }
       // Shutdown the protocol servers in parallel
       protocolServers.values().parallelStream().forEach(ProtocolServer::stop);
       cacheManagers.values().forEach(cm -> SecurityActions.stopCacheManager(cm));
-      status = ComponentStatus.TERMINATED;
+      this.status = ComponentStatus.TERMINATED;
    }
 
-   public void stop() {
-      getExitHandler().exit(0);
+
+   private void serverStopHandler(ExitStatus exitStatus) {
+      scheduler = Executors.newSingleThreadScheduledExecutor();
```

Comment:
we can't use our executor here?

Reply:
Let's see what @wburns thinks.

Reply:
The idea came from @gustavonalle who introduced a delay in single server shutdown mostly as a way to let the server complete the REST request cleanly before initiating shutdown. This doesn't make sense for the actual execution of the shutdown on each node. I'll turn that around.

Reply:
Actually I remember now why I did this:
the server will shutdown *during* execution of the clustered runnable, thus causing an error even though everything was ok.

Reply:
The delay needs to exist to let the caller know the shutdown request was successfully acknowledged and will happen eventually   :)

Reply:
Okay, in that case the only other concern I have is how this will work in Infinispan Quarkus Server. Since we control the whole lifecycle and this is only done while shutting down I am fine with using a new thread.

Reply:
What doesn't GraalVM support this time? :)

Reply:
Nothing with Graal :) Just thinking how the lifecycle of the Quarkus Server works. But we can worry about this later me thinks.